wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+            )
+            with open(
+                os.path.join(context.output_directory, self.preferences_xml + ".new"),
+                "w",
+            ) as write_fh:
+                for line in lines:
+                    write_fh.write(line)
+
+        self.target.push(
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+            "/data/data/{}/shared_prefs/{}".format(
+                self.chrome_package, self.preferences_xml
+            ),
+            as_root=True,
+        )
+
+    def run(self, context):
+        super(Speedometer, self).run(context)
+
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+        self.target.execute(browser_launch_cmd)
+
+        # Wait 60 seconds at least, and then wait until we don't see the 'sandboxed_process' process for 10 seconds.
+        time.sleep(60)
+
+        countdown = 5
+        while countdown > 0:
+            busiest_line = self.target.execute("top -n1 -m1 -q -b").split("\n")[0]
+            while "sandboxed_process" in busiest_line:
+                countdown = 5
+                time.sleep(2)
+                busiest_line = self.target.execute("top -n1 -m1 -q -b").split("\n")[0]
+            time.sleep(2)
+            countdown -= 1
 
     def update_output(self, context):
         super(Speedometer, self).update_output(context)
+
+        self.target.execute("uiautomator dump", as_root=True)
+        self.target.pull(
+            "/sdcard/window_dump.xml", context.output_directory, as_root=True
+        )
+
+        with open(
+            os.path.join(context.output_directory, "window_dump.xml"), "rb"
+        ) as fh:
+            dump = fh.read().decode("utf-8")
+        match = self.regex.search(dump)
         result = None
-        logcat_file = context.get_artifact_path('logcat')
-        with open(logcat_file, errors='replace') as fh:
-            for line in fh:
-                match = self.regex.search(line)
-                if match:
-                    result = float(match.group(1))
+        if match:
+            result = float(match.group(1))
 
         if result is not None:
-            context.add_metric('Speedometer Score', result, 'Runs per minute', lower_is_better=False)
+            context.add_metric(
+                "Speedometer Score", result, "Runs per minute", lower_is_better=False
+            )
         else:
-            raise WorkloadError("The Speedometer workload has failed. No score was obtainable.")
+            raise WorkloadError(
+                "The Speedometer workload has failed. No score was obtainable."
+            )
+
+    def teardown(self, context):
+        super(Speedometer, self).teardown(context)
+        subprocess.check_output(
+            "adb reverse --remove tcp:{}".format(self.webserver_port), shell=True
```

Comment:
Again, `adb` should not be invoked directly -- do not assume it is in path. Should be done via `adb_command`  from devlib utils.